### ISSUE TITLE
Gradient error in backpropagation

### ISSUE THREAD

**Issue by AuroraRyan0301** [graphdeco-inria/gaussian-splatting#551]
https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/cuda_rasterizer/backward.cu#L550
It seems that coeffcient here should be -1 instead of -0.5 from chain rule.
![compute_gaussian_derivative](https://github.com/graphdeco-inria/gaussian-splatting/assets/138105825/e9cd40c1-7ab1-4136-a58c-4698ad3d21a0)

**Comment by KeKsBoTer** [comment]
Hello,

I also discovered this bug/mistake the other day. 
So I cloned the repository, fixed the bug, and rerun all the experiments.
Here are the results (test set, 30000 epochs):

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SSIM</th>
      <th>PSNR</th>
      <th>LPIPS</th>
      <th>SSIM fixed</th>
      <th>PSNR fixed</th>
      <th>LPIPS fixed</th>
      <th>PSRN diff</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>bicycle</th>
      <td>0.762486</td>
      <td>25.170691</td>
      <td>0.216434</td>
      <td>0.753654</td>
      <td>24.906933</td>
      <td>0.224712</td>
      <td>0.263758</td>
    </tr>
    <tr>
      <th>garden</th>
      <td>0.860721</td>
      <td>27.179022</td>
      <td>0.115365</td>
      <td>0.858853</td>
      <td>27.029284</td>
      <td>0.112566</td>
      <td>0.149738</td>
    </tr>
    <tr>
      <th>truck</th>
      <td>0.871009</td>
      <td>24.940348</td>
      <td>0.154585</td>
      <td>0.868983</td>
      <td>24.774864</td>
      <td>0.158739</td>
      <td>0.165483</td>
    </tr>
    <tr>
      <th>room</th>
      <td>0.915746</td>
      <td>31.341377</td>
      <td>0.223303</td>
      <td>0.908076</td>
      <td>30.608786</td>
      <td>0.233915</td>
      <td>0.732592</td>
    </tr>
    <tr>
      <th>treehill</th>
      <td>0.630711</td>
      <td>22.303074</td>
      <td>0.327688</td>
      <td>0.626817</td>
      <td>22.289064</td>
      <td>0.337269</td>
      <td>0.014009</td>
    </tr>
    <tr>
      <th>playroom</th>
      <td>0.900564</td>
      <td>29.925705</td>
      <td>0.244207</td>
      <td>0.897798</td>
      <td>29.634159</td>
      <td>0.254880</td>
      <td>0.291546</td>
    </tr>
    <tr>
      <th>stump</th>
      <td>0.770135</td>
      <td>26.562475</td>
      <td>0.218858</td>
      <td>0.759979</td>
      <td>26.389538</td>
      <td>0.229224</td>
      <td>0.172937</td>
    </tr>
    <tr>
      <th>kitchen</th>
      <td>0.922675</td>
      <td>30.712807</td>
      <td>0.129774</td>
      <td>0.919008</td>
      <td>30.291496</td>
      <td>0.135454</td>
      <td>0.421310</td>
    </tr>
    <tr>
      <th>drjohnson</th>
      <td>0.896235</td>
      <td>28.938360</td>
      <td>0.247888</td>
      <td>0.895815</td>
      <td>28.829479</td>
      <td>0.257125</td>
      <td>0.108881</td>
    </tr>
    <tr>
      <th>flowers</th>
      <td>0.601902</td>
      <td>21.447706</td>
      <td>0.340630</td>
      <td>0.382011</td>
      <td>15.328293</td>
      <td>0.564916</td>
      <td>6.119413</td>
    </tr>
    <tr>
      <th>bonsai</th>
      <td>0.937666</td>
      <td>31.979340</td>
      <td>0.208291</td>
      <td>0.931681</td>
      <td>31.220346</td>
      <td>0.217860</td>
      <td>0.758993</td>
    </tr>
    <tr>
      <th>counter</th>
      <td>0.904743</td>
      <td>28.888483</td>
      <td>0.204061</td>
      <td>0.893800</td>
      <td>28.274176</td>
      <td>0.218922</td>
      <td>0.614307</td>
    </tr>
    <tr>
      <th>train</th>
      <td>0.804521</td>
      <td>21.770004</td>
      <td>0.217442</td>
      <td>0.793224</td>
      <td>20.911655</td>
      <td>0.232966</td>
      <td>0.858349</td>
    </tr>
  </tbody>
</table>


It seems like the reconstructions are consistently better with the incorrect gradient. 
For the flower scene the reconstruction even "fails" and results in a blurry mess: 
![00003](https://github.com/graphdeco-inria/gaussian-splatting/assets/14186588/750e02a6-62b1-4391-b859-d60e34362f02)

This is a very strange behavior that needs more investigation...

**Comment by KeKsBoTer** [comment]
I think I just found out why the gradients are correct. 
Later in the backward pass ([here](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/cuda_rasterizer/backward.cu#L210C1-L210C104)) the gradient is multiplied by 2.

**Comment by kwea123** [comment]
yes, they multiply `2` to all the terms with `dL_dconic.y` on https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/cuda_rasterizer/backward.cu#L210-L212
This is extremely confusing but it is correct

**Comment by AuroraRyan0301** [comment]
> I think I just found out why the gradients are correct. Later in the backward pass ([here](https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/cuda_rasterizer/backward.cu#L210C1-L210C104)) the gradient is multiplied by 2.

Thx, you are right, I understand it now. I also did similar experiment and found PSNR decreased if I changed the coeff from -0.5 to -1. The code is kind of weired.
